### PR TITLE
db connectivity

### PR DESCRIPTION
**Pull request by realwebdev**
gorm structs

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,6 @@
+DIALECT="postgres"
+HOST="localhost"
+DBPORT="5432"
+USER="postgres"
+NAME="clockify3"
+PASSWORD="password"
```
Comment (Bilaltanveer38) [inline .env:6]: .env file should not be included in the git repo, must add .gitignore file

File: main.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+type Activity struct {
+	activity_id uint `gorm:"primaryKey"`
+	project_id  uint
+	user_id     uint
+	start_time  time.Time
+	end_time    time.Time
+	total_time  time.Time
+}
+
```
Comment (Bilaltanveer38) [inline main.go:45]: There should be a package named models, and each struct have its own file.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	// Loading enviroment variables
+	dialect := os.Getenv("DIALECT")
+	host := os.Getenv("HOST")
+	dbPort := os.Getenv("DBPORT")
+	user := os.Getenv("USER")
+	dbname := os.Getenv("NAME")
+	dbpassword := os.Getenv("PASSWORD")
+
```
Comment (Bilaltanveer38) [inline main.go:64]: There should be a seperate package conf with conf.go file, and all the handling of environment variables should be done there.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	user_id     uint
+	start_time  time.Time
+	end_time    time.Time
+	total_time  time.Time
+}
+
+var (
+	user = &User{user_id: 1, username: "haseeb", email: "haseeb@shalan@sasketchwan.com", password: "password"}
+)
+
```
Comment (Bilaltanveer38) [inline main.go:49]: Write a function CreateUser(username, email, password string) instead of making a variable

File: main.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/joho/godotenv"
+
+	"github.com/jinzhu/gorm"
+
+	_ "github.com/jinzhu/gorm/dialects/postgres"
+)
+
+var db *gorm.DB
+var err error
+
```
Comment (Bilaltanveer38) [inline main.go:18]: We should never create the global variables

File: main.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		fmt.Println("Connected to database successfully")
+	}
+
+	defer db.Close()
+
+	db.AutoMigrate(&User{})
+	db.AutoMigrate(&Project{})
+	db.AutoMigrate(&Activity{})
+
+	db.Create(user)
```
Comment (Bilaltanveer38) [inline main.go:83]: This should have handled in CreateUser function

File: main.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	user := os.Getenv("USER")
+	dbname := os.Getenv("NAME")
+	dbpassword := os.Getenv("PASSWORD")
+
+	// Database connection string
+	dbURI := fmt.Sprintf("host=%s user=%s dbname=%s sslmode=disable password=%s port=%s", host, user, dbname, dbpassword, dbPort)
+
+	// Openning connection to database
+	db, err = gorm.Open(dialect, dbURI)
+
```
Comment (Bilaltanveer38) [inline main.go:70]: There should be a package named Datastore or DB, and all of the db based operations like establishing connection returning DB object, should be done there.

File: main.go (hunk 6 above)
Comment (Bilaltanveer38) [reply main.go:83]: Moreover, the project is not 'gofmt', make sure to gofmt the project before pushing it on the repo.

**Review by Bilaltanveer38 — CHANGES_REQUESTED** [review]

File: database/conn.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/realwebdev/Bilal/clockify3/models"
+)
+
+func ConnectDB() *gorm.DB {
+
+	err := godotenv.Load()
+	if err != nil {
+		log.Fatal("Error loading .env file")
+	}
+
```
Comment (zunnorainzaheer) [inline database/conn.go:20]: Should load the env file in the separate package named conf

Comment (Bilaltanveer38) [inline database/conn.go:20]: Should load the env file in the separate package named conf

File: database/conn.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	dialect := os.Getenv("DIALECT")
+	host := os.Getenv("HOST")
+	dbPort := os.Getenv("DBPORT")
+	user := os.Getenv("USER")
+	dbname := os.Getenv("NAME")
+	dbpassword := os.Getenv("PASSWORD")
+
+	dbURI := fmt.Sprintf("host=%s user=%s dbname=%s sslmode=disable password=%s port=%s", host, user, dbname, dbpassword, dbPort)
+	var errr error
```
Comment (Bilaltanveer38) [inline database/conn.go:29]: use err instead errr

File: database/conn.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	var db *gorm.DB
+	// Openning connection to database
+	db, errr = gorm.Open(dialect, dbURI)
```
Comment (Bilaltanveer38) [inline database/conn.go:32]: or use this db,err:= gorm.Open(dialect, dbURI) 
this syntax is recommended

File: database/conn.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+	if errr != nil {
+		log.Fatal(errr)
+	} else {
+		fmt.Println("Connected to database successfully")
+	}
+
```
Comment (Bilaltanveer38) [inline database/conn.go:39]: return db

File: database/conn.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+	dbURI := fmt.Sprintf("host=%s user=%s dbname=%s sslmode=disable password=%s port=%s", host, user, dbname, dbpassword, dbPort)
+	var errr error
+	var db *gorm.DB
+	// Openning connection to database
+	db, errr = gorm.Open(dialect, dbURI)
+
+	if errr != nil {
+		log.Fatal(errr)
+	} else {
+		fmt.Println("Connected to database successfully")
```
Comment (Bilaltanveer38) [inline database/conn.go:37]: remove else block

File: database/conn.go [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	defer db.Close()
+	var (
+		user1 = &models.User{user_id: 1, username: "haseeb", email: "haseeb@shalan@sasketchwan.com", password: "password"}
+	)
+	// Loading enviroment variables
```
Comment (Bilaltanveer38) [inline database/conn.go:44]: no need to create the object

File: database/conn.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+	db.AutoMigrate(&models.User{})
+	db.AutoMigrate(&models.Project{})
+	db.AutoMigrate(&models.Activity{})
+
```
Comment (Bilaltanveer38) [inline database/conn.go:49]: create a wrapper function instead of calling this

**Pull request merged**